freedomofpress/securedrop
# Makes unattended-upgrades sequencing more similar to cron-apt

## Status

Ready for review 

## Description of Changes

Fixes #5851. Fixes #5849. 

Builds on the great work in #5852 & #5853. 

* Overrides the systemd "timer" declarations for apt update & apt upgrade.
* Modifies the unattended-upgrades desired reboot time to be "now", i.e. as soon as the upgrade is finished
* Makes timing of upgrade and reboots handled via Ansible
* Updates test logic to use `apt-config`, to avoid problems like #5852 in the future

## Testing

1. Configure a local staging environment on this branch, including using latest debs built from here.
2. To observe unattended-upgrades in action, pick a new time for `daily_reboot_time` 1-2h in the future. Re-run the config of just those files to apply, like so: `molecule converge -s qubes-staging-focal -- --tags ua -e daily_reboot_time=15` 
3. Make sure to `apt-mark unhold securedrop-app-code`, or even `apt-mark showhold | xargs -r sudo apt-mark unhold` so that ua can make changes. Also run `rm -f /var/lib/apt/periodic/{upgrade,update}-stamp` so that ua doesn't defer updates based on recent changes.
4. Run `apt-cache policy securedrop-app-code` and confirm its outdated. 
5. Wait. Optionally, procure coffee or tea. 
6. Check back and confirm that packages have actually updated, particularly that `securedrop-app-code` is at least `1.8.0~rc3`. 
7. Finally, test the minus-one-hour cron timing logic and confirm you see different values in the `apt-daily.timer` and `apt-daily-upgrade.timer`, different by one hour. For example, for `molecule converge -s qubes-staging-focal -- --tags ua -e daily_reboot_time=0` you should see a value of midnight for the upgrade, but 11PM (23:00) for the apt update.


## Deployment
Critical for Focal behavior starting in in 1.8.0.
